File: 2020/day-01/solutions/day01.rb
numbers = ARGF.map{ |line| line.to_i(10) }

part1 = nil
part2 = nil

numbers.each do |i|
  numbers.each do |j|
    if part1 == nil and i + j == 2020
      part1 = i*j
    end

    numbers.each do |k|
      if part2 == nil and i + j + k == 2020
        part2 = i*j*k
      end
    end
  end
end

puts part1
puts part2

File: 2020/day-01/solutions/day01.rs
use std::{io, io::prelude::*};

fn main() {
  let stdin = io::stdin();
  let lines = stdin.lock().lines().filter_map(|line| line.ok()?.parse::<i64>().ok()).collect::<Vec<i64>>();
  let lines = &lines;

  let triplets = lines.iter().cloned().flat_map(|i|
      lines.iter().cloned().flat_map(move |j|
        lines.iter().copied().map(move |k| (i, j, k)))).collect::<Vec<(i64,i64,i64)>>();
  let triplets = &triplets;

  let part1 = triplets.iter().find_map(|(i,j,_k)| if i+j == 2020 { Some(i*j) } else { None } );
  let part2 = triplets.iter().find_map(|(i,j,k)| if i+j+k == 2020 { Some(i*j*k) } else { None } );

  println!("{}", part1.unwrap_or_default());
  println!("{}", part2.unwrap_or_default());
}
